src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,250 @@
+    if(stencilInstantiation->getOptimizerContext()->getOptions().ReportBoundaryConditions) {
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName() << " :";
+      std::cout << " No boundary conditions applied\n";
+    }
+    return true;
+  }
+  // returns the original ID of a variable since we need to make sure that if
+  auto getOriginalID = [&](int ID) {
+    // check if the variable exists to be sure not to assert
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(
+        stencilInstantiation->getOriginalNameFromAccessID(ID));
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      if(stencilInstantiation->isField(ID)) {
+        return stencilInstantiation->getAccessIDFromName(
+            stencilInstantiation->getOriginalNameFromAccessID(ID));
+      } else {
+        return -1;
+      }
+    } else {
+      return -1;
+    }
+
+  };
+
+  std::unordered_map<int, Extents> dirtyFields;
+  std::unordered_map<int, bcStorage> allBCs;
+
+  // Fetch all the boundary conditions stored in the instantiation
+  for(const auto& bc : stencilInstantiation->getBoundaryConditions()) {
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(bc.first);
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      int ID = stencilInstantiation->getAccessIDFromName(bc.first);
+      bcStorage storage;
+      storage.functor = bc.second.functor;
+      for(int i = 0; i < bc.second.arguments.size(); ++i) {
+        storage.arguments.push_back(bc.second.arguments[i]);
+      }
+      allBCs.emplace(ID, storage);
+    }
+  }
+
+  // Loop through all the StmtAccessPair in the stencil forward
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+    Stencil& stencil = *stencilPtr;
+    DAWN_LOG(INFO) << "analyzing stencil " << stencilInstantiation->getName();
+    std::unordered_map<int, Extents> stencilDirtyFields;
+    stencilDirtyFields.clear();
+
+    for(const auto& multiStagePtr : stencil.getMultiStages()) {
+      MultiStage& multiStage = *multiStagePtr;
+      for(auto stageIt = multiStage.getStages().begin(); stageIt != multiStage.getStages().end();
+          ++stageIt) {
+        Stage& stage = (**stageIt);
+        for(const auto& domethod : stage.getDoMethods()) {
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
```

Comment:
early return `if(originalID== -1) continue;` ?

Reply:
we can move this up one call

Reply:
uhmm where? I didnt understand it